### PR TITLE
feat(events)!: improved event bus system

### PR DESCRIPTION
**Pull request by joshuasing**
**Summary**
Replace existing event system with a faster event bus system.

**Changes**
 - Update README.md to use Sonatype repository
 - Remove existing event system
 - Add new event bus system
 - Add event bus tests
 - Bump version to `0.10.0-SNAPSHOT`
 - Add event subscriber builder
 - Make `UserConnectEvent` cancellable
 - Add `ServerUserKickEvent`
 - Add `User#getAddress()`
 - Add `User#chat(Component)`
 - Add `User#disconnect(Component)`
 - Add catch for chat modification on Velocity for Minecraft 1.19.1+ players

**Checklist**
 - [x] I acknowledge and agree to the [Developer Certificate of Origin](https://developercertificate.org/).
 - [x] All contributed code can be distributed under the terms of the [MIT License](https://github.com/ChameleonFramework/Chameleon/blob/main/LICENSE).
 - [x] I have checked the ["Allow edit from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) option.
 - [ ] I have added appropriate unit tests for my changes. 

**This pull request contains breaking changes.**

### CODE REVIEW

File: api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
 /**
  * {@link User} connect event, dispatched when a user joins the proxy/server.
  */
-public class UserConnectEvent implements UserEvent {
+public final class UserConnectEvent implements UserEvent {
```
Comment (SLLCoding) [inline api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java:32]: Should probably allow cancellation on this event.

File: api/src/main/java/dev/hypera/chameleon/events/proxy/ProxyUserSwitchEvent.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
  * {@link ProxyUser} switch sever event, dispatched whenever a player switches server.
  */
 @PlatformSpecific(Platform.Type.PROXY)
-public class ProxyUserSwitchEvent implements ProxyUserEvent {
+public final class ProxyUserSwitchEvent implements ProxyUserEvent {
```
Comment (SLLCoding) [inline api/src/main/java/dev/hypera/chameleon/events/proxy/ProxyUserSwitchEvent.java:38]: This event should also probably be cancellable.

**Review by SLLCoding — DISMISSED** [review]

File: api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java (hunk 1 above)
Comment (joshuasing) [reply api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java:32]: This event is not cancellable on any platform.
To do this, we would need to implement kick-related events and kick the player when the event is cancelled.

File: api/src/main/java/dev/hypera/chameleon/events/proxy/ProxyUserSwitchEvent.java (hunk 2 above)
Comment (joshuasing) [reply api/src/main/java/dev/hypera/chameleon/events/proxy/ProxyUserSwitchEvent.java:38]: This event is not cancellable on BungeeCord or Velocity unless we listen to the pre-switch events which could be too early for dispatching this event.

File: api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java (hunk 1 above)
Comment (SLLCoding) [reply api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java:32]: I believe it'll be worth implementing that.

Comment (joshuasing) [reply api/src/main/java/dev/hypera/chameleon/events/common/UserConnectEvent.java:32]: Should that be done in this pull request or a separate one?

File: platform-velocity/src/main/java/dev/hypera/chameleon/platform/velocity/events/VelocityListener.java [hunk 3]
```diff
@@ -90,24 +101,33 @@ public void onChatEvent(@NotNull PlayerChatEvent event) {
...
+
+    /**
+     * Platform {@link UserDisconnectEvent} handler.
+     *
+     * @param event Platform event.
+     */
+    @Subscribe
+    public void onKickedFromServerEvent(@NotNull KickedFromServerEvent event) {
+        this.chameleon.getEventBus().dispatch(new UserDisconnectEvent(wrap(event.getPlayer()), event.getServerKickReason().map(AdventureConverter::convertComponentBack).orElse(null)));
     }
```
Comment (SLLCoding) [inline platform-velocity/src/main/java/dev/hypera/chameleon/platform/velocity/events/VelocityListener.java:119]: This technically isn't a disconnect?

File: api/src/main/java/dev/hypera/chameleon/events/common/UserDisconnectEvent.java [hunk 4]
```diff
@@ -20,27 +20,33 @@
...
+import org.jetbrains.annotations.Nullable;
 
 /**
  * {@link User} disconnect event, dispatched when a user leaves the proxy/server.
  */
-public class UserDisconnectEvent implements UserEvent {
+public final class UserDisconnectEvent implements UserEvent {
 
     private final @NotNull User user;
+    private final @Nullable Component reason;
```
Comment (SLLCoding) [inline api/src/main/java/dev/hypera/chameleon/events/common/UserDisconnectEvent.java:38]: This should probably be `@NotNull` with a default value.

File: api/src/main/java/dev/hypera/chameleon/events/common/UserDisconnectEvent.java [hunk 5]
```diff
@@ -52,4 +58,13 @@ public UserDisconnectEvent(@NotNull User user) {
...
     }
 
+    /**
+     * Get the reason for this disconnection.
+     *
+     * @return disconnect reason, if available, otherwise empty.
+     */
+    public @NotNull Optional<Component> getReason() {
+        return Optional.ofNullable(this.reason);
+    }
```
Comment (SLLCoding) [inline api/src/main/java/dev/hypera/chameleon/events/common/UserDisconnectEvent.java:68]: If the field becomes `@NotNull`, then this should not be an `Optional`.

**Review by SLLCoding — CHANGES_REQUESTED** [review]

File: platform-velocity/src/main/java/dev/hypera/chameleon/platform/velocity/events/VelocityListener.java [hunk 6]
```diff
@@ -63,51 +64,70 @@ public VelocityListener(@NotNull VelocityChameleon chameleon) {
...
 
         if (chameleonEvent.isCancelled()) {
-            event.setResult(ChatResult.denied());
+            if (event.getPlayer().getProtocolVersion().getProtocol() >= 760) {
+                this.chameleon.getInternalLogger().error("Failed to cancel a chat message, doing so would result in Velocity throwing an exception and the sender being kicked.");
+                this.chameleon.getInternalLogger().error("This IS NOT a bug, but rather an intentional change in Velocity caused by changes in Minecraft 1.19.1.");
+                this.chameleon.getInternalLogger().error("See https://github.com/PaperMC/Velocity/issues/804 for more information.");
+            } else {
+                event.setResult(ChatResult.denied());
+            }
```
Comment (SLLCoding) [inline platform-velocity/src/main/java/dev/hypera/chameleon/platform/velocity/events/VelocityListener.java:107]: Should probably be put into a private method to make changes lightly easier. Just a nit-pick.

**Review by SLLCoding — APPROVED** [review]

**Review by SLLCoding — DISMISSED** [review]

**Review by SLLCoding — APPROVED** [review]

**Pull request merged**